sys/cheri/cheri_cap_relocs.c

```diff
@@ -60,6 +64,37 @@ init_linker_file_cap_relocs(const void *start_relocs, const void *stop_relocs,
 #else
 	bool can_set_code_bounds = true;
 #endif
-	cheri_init_globals_impl(start_relocs, stop_relocs, data_cap, code_cap,
-	    data_cap, can_set_code_bounds, base_addr);
+
+	/*
+	 * This cannot use cheri_init_globals_impl directly as symbols
+	 * for kernel modules in the vnet and dpcpu sets need to use
+	 * alternate base addresses and capabilities.  Instead, we
+	 * invoke a caller-supplied callback on each capability to
+	 * request the base address and source capability.
+	 */
+	for (const struct capreloc *reloc = start_relocs;
```

Comment:
While this is not a lot of code duplication, I'd prefer if we could make the compiler-provided header more flexible. I think adding it here is fine as long as we have an issue in llvm-project tracking this missing feature.

Reply:
I had talked about that with Jess on IRC a bit.  I think one could make the "core" function in the header take a callback and have the existing functions construct a struct with the three caps (pc, data, rodata) and use a default callback that uses that struct for existing API functions.